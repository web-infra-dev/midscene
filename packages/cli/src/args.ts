export type ArgumentValueType = string | boolean | number;

export interface Argument {
  name: string;
  value: ArgumentValueType;
}

export function parse(args: string[]): Argument[] {
  const orderedArgs: Argument[] = [];
  args.forEach((arg, index) => {
    if (arg.startsWith('--')) {
      const key = arg.substring(2);
      let value: ArgumentValueType =
        args[index + 1] && !args[index + 1].startsWith('--')
          ? args[index + 1]
          : true;

      if (typeof value === 'string' && /^\d+$/.test(value)) {
        value = Number.parseInt(value, 10);
      }
      orderedArgs.push({ name: key, value });
    }
  });

  return orderedArgs;
}

export function findOnlyItemInArgs(
  args: Argument[],
  name: string,
): ArgumentValueType {
  const found = args.filter((arg) => arg.name === name);
  if (found.length === 0) {
    return false;
  }

  if (found.length > 1) {
    throw new Error(`Multiple values found for ${name}`);
  }

  return found[0].value;
}
