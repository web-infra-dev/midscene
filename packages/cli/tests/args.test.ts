import { findOnlyItemInArgs, orderMattersParse } from '@/args';
import { describe, expect, test } from 'vitest';

describe('args', () => {
  test('should parse arguments', async () => {
    expect(
      findOnlyItemInArgs({ url: 'https://example.com', _: [] }, 'url'),
    ).toBe('https://example.com');
    expect(() => {
      findOnlyItemInArgs(
        { url: 'https://example.com', _: [], query: [1, 2] },
        'query',
      );
    }).toThrowError('Multiple values found for query');
  });

  test('should ignore the node script name', async () => {
    const input = [
      'node',
      '/path/to/bin.js',
      '--url',
      'https://example.com',
      '--action',
      '--sleep',
      '20',
      '--sleep',
      '10',
    ];

    const result = orderMattersParse(input);

    expect(result).toEqual([
      {
        name: 'url',
        value: 'https://example.com',
      },
      {
        name: 'action',
        value: true,
      },
      {
        name: 'sleep',
        value: 20,
      },
      {
        name: 'sleep',
        value: 10,
      },
    ]);
  });
});
