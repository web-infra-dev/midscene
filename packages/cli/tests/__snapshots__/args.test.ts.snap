// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`args > should parse arguments 1`] = `
[
  {
    "name": "url",
    "value": "https://example.com",
  },
  {
    "name": "width",
    "value": 500,
  },
  {
    "name": "action",
    "value": "click",
  },
  {
    "name": "assert",
    "value": "this is an assertion",
  },
  {
    "name": "query-output",
    "value": "output.json",
  },
  {
    "name": "query",
    "value": "title",
  },
  {
    "name": "query",
    "value": "content",
  },
  {
    "name": "prefer-cache",
    "value": true,
  },
]
`;
